### PR TITLE
Unmodifiable collection view

### PR DESCRIPTION
**Pull request by Sharom**
This pull-request adds class `UnmodifiableCollectionView` to be able to return a collection for read-only.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,5 @@
 vendor/
 composer.lock
 build/
+.idea
+*.iml
```
Comment (stof) [inline .gitignore]: -1 for that. these are not project-specific but system-specific. you should use a system-level ignore rule: https://help.github.com/articles/ignoring-files/

File: lib/Doctrine/Common/Collections/UnmodifiableCollectionView.php [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+    {
+        return $this->collection->count();
+    }
+
+    /**
+     * @param string $message
+     *
+     * @return BadMethodCallException
+     */
+    private function createUnsupportedException($message)
```
Comment (stof) [inline lib/Doctrine/Common/Collections/UnmodifiableCollectionView.php]: IMO, you don't need this private method

Comment (Sharom) [inline lib/Doctrine/Common/Collections/UnmodifiableCollectionView.php]: agree

File: .gitignore (hunk 1 above)
Comment (Sharom) [inline .gitignore]: fixed

File: lib/Doctrine/Common/Collections/UnmodifiableCollectionView.php [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+use BadMethodCallException;
+
+/**
+ * Unmodifiable view of a collection.
+ *
+ * Wraps original collection and protects it from unplanned change.
+ * 
+ * @author Roman Shamritskiy <roman@litgroup.ru>
+ */
+class UnmodifiableCollectionView implements Collection
```
Comment (stof) [inline lib/Doctrine/Common/Collections/UnmodifiableCollectionView.php]: would be great to implement Selectable too

Comment (Sharom) [inline lib/Doctrine/Common/Collections/UnmodifiableCollectionView.php]: > would be great to implement Selectable too

I thought about it. But I still not sure that it's good solution.
What if a wrapped collection does not implement `Selectable`?
To throw `BadMethodCallException`? It's not clear.

**Comment by Ocramius** [comment]
A collection that exposes not-implemented methods that throw exceptions seems wrong to me.

You now disallow methods that are expected to work (without failure) on the consumer side.

While I like where you are going with this, this is not the way to implement it. The correct implementation would be to segregate the read operations in own interfaces, and then implement only those.

I wouldn't do that though, as everyone currently relies on the `Collection` interface directly, so some more in-depth analysis of the problem and how we can evolve the library has to be done.

Closing as `invalid` due to the remarks above.

**Comment by fprochazka** [comment]
@Sharom if you haven't already created a package, you might wanna look into this https://github.com/Kdyby/DoctrineCollectionsReadonly  :)

**Comment by Sharom** [comment]
> The correct implementation would be to segregate the read operations in own interfaces, and then implement only those.

@Ocramius you are right. But... we already have a `Collection`. And I still want be able to return read-only collection from an entity, and prevent unexpected mutation of a collection from outside of aggregate.

Doctrine documentation contains one example:

```php
public function getReadComments() {
    return $this->commentsRead->toArray();
}
```

but this is wrong way. What about a lazy collection?

Implementation of  `UnmodifiableListView` is not what I invented yesterday (for example: https://api.dartlang.org/stable/1.16.0/dart-collection/UnmodifiableListView-class.html)

`BadMethodCallException` is a subtype of `RuntimeException` which evidences of breach of contract.
And this solution works quite successfully. You can think about immutability of a collection as about one of the states of a collection.

Segregation of interfaces for reading/changing of a collection may be a good solution, but it may be not. I didn't ever see implementation of generic collections with segregation of interfaces for reading and for modification.

`return $this->commentsRead->toArray()` – this is not a solution. If `UnmodifiableListView` not a solution too, let's think about necessity of a read-only collection together.

P.S. (If any sentences incorrect in the context of english, I am sorry about it.
I learn English ;))

**Comment by Sharom** [comment]
@fprochazka

> if you haven't already created a package, you might wanna look into this https://github.com/Kdyby/DoctrineCollectionsReadonly :)

Thanks 👍

**Pull request closed without merging**